loader/loader.c

```diff
@@ -5912,52 +5913,102 @@ VkResult setup_loader_term_phys_devs(struct loader_instance *inst) {
         loader_log(inst, VULKAN_LOADER_ERROR_BIT, 0,
-                   "setup_loader_term_phys_devs:  Failed to allocate new physical device array of size %d", inst->total_gpu_count);
+                   "setup_loader_term_phys_devs:  Failed to allocate new physical device array of size %d", new_phys_devs_count);
         res = VK_ERROR_OUT_OF_HOST_MEMORY;
         goto out;
     }
-    memset(new_phys_devs, 0, sizeof(struct loader_physical_device_term *) * inst->total_gpu_count);
+    memset(new_phys_devs, 0, sizeof(struct loader_physical_device_term *) * new_phys_devs_count);
+
+    // Current index into the new_phys_devs array - increment whenever we've written in.
+    uint32_t idx = 0;
+
+    // Copy over everything found through sorted enumeration
+    for (uint32_t i = 0; i < windows_sorted_devices_count; ++i) {
+        for (uint32_t j = 0; j < windows_sorted_devices_array[i].device_count; ++j) {
+            // Check if this physical device is already in the old buffer
+            if (NULL != inst->phys_devs_term) {
+                for (uint32_t old_idx = 0; old_idx < inst->phys_dev_count_term; old_idx++) {
+                    if (windows_sorted_devices_array[i].physical_devices[j] == inst->phys_devs_term[old_idx]->phys_dev) {
+                        new_phys_devs[idx] = inst->phys_devs_term[old_idx];
+                        break;
+                    }
+                }
+            }
 
+            // If this physical device isn't in the old buffer, then we need to create it.
+            if (NULL == new_phys_devs[idx]) {
+                new_phys_devs[idx] = loader_instance_heap_alloc(inst, sizeof(struct loader_physical_device_term),
+                                                                VK_SYSTEM_ALLOCATION_SCOPE_INSTANCE);
+                if (NULL == new_phys_devs[idx]) {
+                    loader_log(inst, VULKAN_LOADER_ERROR_BIT, 0,
+                               "setup_loader_term_phys_devs:  Failed to allocate physical device terminator object %d", idx);
+                    inst->total_gpu_count = idx;
+                    res = VK_ERROR_OUT_OF_HOST_MEMORY;
+                    goto out;
+                }
+
+                loader_set_dispatch((void *)new_phys_devs[idx], inst->disp);
+                new_phys_devs[idx]->this_icd_term = windows_sorted_devices_array[i].icd_term;
+                new_phys_devs[idx]->icd_index = (uint8_t)(windows_sorted_devices_array[i].icd_index);
+                new_phys_devs[idx]->phys_dev = windows_sorted_devices_array[i].physical_devices[j];
+            }
+
+            // Increment the count of new physical devices
+            idx++;
+        }
+    }
+
+    // Now go through the rest of the physical devices and add them to new_phys_devs
 #ifdef LOADER_ENABLE_LINUX_SORT
     if (is_linux_sort_enabled(inst)) {
-        for (uint32_t dev = 0; dev < inst->total_gpu_count; ++dev) {
+        for (uint32_t dev = idx; dev < new_phys_devs_count; ++dev) {
```

Comment:
When would `idx` be non-0 since the above code only runs for Windows sorted items?

Reply:
The code path runs everytime we have a driver that doesn't support Version 6. This is the crux of the issue reported in the PR this PR is replacing. I had a hard time reproducing it because the internal structures weren't filled out, while the application visible `VkPhysicalDevice` was non-null. In other words, the `struct loader_physical_device_term` array was only partially filled in.

Reply:
But this code is #ifdef'd in LOADER_ENABLE_LINUX_SORT so it is not run on Windows.

Reply:
While I do see that now, I'd rather not have the same issue again when someone enables LOADER_ENABLE_LINUX_SORT on windows platforms, which should be possible due to how the linux sorting doesn't rely on any linux specific functions.